keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,739 @@
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Performance
+Using namespace selector will make it easier for users to specify affinity 
+constraints across a large number of namespaces. The initial implementation 
+of pod affinity/anti-affinity suffered from performance challenges, however 
+over the releases 1.17 - 1.20 we significantly improved that. 
+
+We currently have integration and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+#### Security/Abuse
+See previous discussion [here](https://github.com/kubernetes/kubernetes/issues/68827#issuecomment-468470598)
+NamespaceSelector will allow users to select all namespaces. This may cause a security 
+concern: a pod with anti-affinity constraint can block pods from all other namespaces 
+from getting scheduled in a failure domain.
+
+We will address this concern by introducing a new quota scope named `CrossNamespaceAffinity`
+that admins can use to limit which namespaces are allowed to have pods with affinity terms
+that set the existing `namespaces` field or the proposed one `namespaceSelector`. 
+
+Using this new scope, admins can prevent certain namespaces from having pods that use
+cross-namespace pod affinity by creating a resource quota object in that namespace with
+`CrossNamespaceAffinity` scope and hard limit of 0:
+
+```go
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+    name: pods-medium
+spec:
+    hard:
+      pods: "0"
+    scopeSelector:
+      matchExpressions:
+      - scopeName: CrossNamespaceAffinity
+```
+
+
+If admins want to disallow using `namespaces` and `namespaceSelector` by default, and 
+only allow it for specific namespaces,  they could configure `CrossNamespaceAffinity` 
+as a limited resource by setting the kube-apiserver flag --admission-control-config-file
+to the path of the following configuration file:
+
+```go
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: "ResourceQuota"
+  configuration:
+    apiVersion: apiserver.config.k8s.io/v1
+    kind: ResourceQuotaConfiguration
+    limitedResources:
+    - resource: pods
+      matchScopes:
+      - scopeName: CrossNamespaceAffinity
+```
+
+With the above configuration, pods can use `namespaces` and `namespaceSelector` only
+if the namespace where they are created have a resource quota object with 
+`CrossNamespaceAffinity` scope and a hard limit equal to the number of pods that are
+allowed to.
+
+Moreover, to prevent accidentally selecting a large number of namespaces, we will reject empty
```

Comment:
@liggitt I am inclined to drop this requirement since we have the quota scope. This will make it harder to automate 1:1 pod placement.

Reply:
What exactly are you proposing to change? I think only allowing `In` is good. But the value of the label could be an empty string.

Reply:
Proposing to have no constraints on NamespaceSelector, just like LabelSelector. While restricting to `In` could help preventing accidental uses of the field, it means namespaces should have a common label, which may be tricky to achieve in some setups (e.g. a provider trying to automate 1:1 pod to node placement).

Reply:
So you are saying you want to support "match any namespace"

Reply:
I think a provider would at least want to have "kube-system" out of the affected namespaces

Reply:
Users could say "NotIn" kube-system (https://github.com/kubernetes/enhancements/pull/2162 will add a label for all namespaces that represent the namespace name).

Reply:
I guess we could still disallow empty selector to avoid accidental uses of the field. NotIn should suffice.

Reply:
I updated the KEP that we will only disallow empty selector to avoid accidental selection of all namespaces. Other operators are allowed.

Reply:
The KEP still says: `The only valid operator is In.` is that the intention?

Reply:
This paragraph is correct. I updated the comment on the field.